### PR TITLE
Api 33

### PR DESCRIPTION
**Pull request by PatDyn**
Refactoring to request Notification Permission after Privacy agreement:

* Move DataStoreModule to own package 
* Introduce Generic DataStore Interface
* Introduce DataStoreRepository
* Introduce ExodusConfig Type
* Move main initialization to dialog fragment
* Ask permissions depending on API Level
* Implement tests

Due to major changes in init flow the following had to be done:
* Fix edge network availability edge cases 
* Add swipe refresh, when only shimmer layout is shown

General Bug Fix:
* Implement try catch cases for when activities are not available

### CODE REVIEW

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/dialog/ExodusDialogFragment.kt [hunk 1]
```diff
@@ -1,46 +1,94 @@
...
+            }.setNegativeButton(getString(R.string.reject)) { _, _ ->
                 activity?.finish()
+            }.create()
+    }
+
+    private val requestPermissionLauncher = registerForActivityResult(
+        ActivityResultContracts.RequestPermission()
+    ) { isGranted: Boolean ->
+        if (isGranted) {
+            exodusDialogViewModel.saveNotificationPermissionRequested(true)
```
Comment (Jean-BaptisteC) [inline app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/dialog/ExodusDialogFragment.kt:65]: Why use condition to execute same method ?

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/trackers/TrackersFragment.kt [hunk 2]
```diff
@@ -56,7 +57,21 @@ class TrackersFragment : Fragment(R.layout.fragment_trackers) {
                 binding.shimmerLayout.visibility = View.GONE
                 trackersRVAdapter.submitList(sortedList)
             } else {
+                binding.swipeRefreshLayout.visibility = View.VISIBLE
                 binding.shimmerLayout.visibility = View.VISIBLE
+                Snackbar.make(
+                    binding.shimmerLayout,
+                    R.string.refresh_needed,
+                    Snackbar.LENGTH_INDEFINITE
```
Comment (Jean-BaptisteC) [inline app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/trackers/TrackersFragment.kt:65]: Doing same thing as appFragment

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 13
✅ Tests

Some comments

**Comment by Jean-BaptisteC** [comment]
❌ ExodusUpdateServiceTest Failed
```
java.lang.NullPointerException: serviceRule.bindService(serviceIntent) must not be null
at org.eu.exodus_privacy.exodusprivacy.ExodusUpdateServiceTest.countsAppsHavingTrackersCorrectly(ExodusUpdateServiceTest.kt:82)
```

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt [hunk 3]
```diff
@@ -79,7 +80,21 @@ class AppsFragment : Fragment(R.layout.fragment_apps) {
                 binding.shimmerLayout.visibility = View.GONE
                 appsRVAdapter.submitList(it)
             } else {
+                binding.swipeRefreshLayout.visibility = View.VISIBLE
                 binding.shimmerLayout.visibility = View.VISIBLE
+                Snackbar.make(
```
[the remark this answers is not in the document]
Comment (PatDyn) [reply app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt]: Ok, i removed the snackbar code. It got re-added when pulling. 
It might be just enough to have the swipeRefreshLayout visible to be able to refresh at any time.

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/dialog/ExodusDialogFragment.kt (hunk 1 above)
Comment (PatDyn) [reply app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/dialog/ExodusDialogFragment.kt]: Good point. I changed that.

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/MainActivityViewModel.kt [hunk 4]
```diff
@@ -1,23 +1,57 @@
...
+        Log.d(TAG, "Got status for notification permission: $status.")
+        val newConfig = config.value as MutableMap
+        newConfig["notification_perm"] = ExodusConfig("notification_requested", status)
+        viewModelScope.launch {
+            Log.d(TAG, "Saving new config: $newConfig.")
+            configStorage.insert(newConfig)
+        }
+    }
+
+    fun saveAppSetup(status: Boolean) {
```
[the remark this answers is not in the document]
Comment (PatDyn) [reply app/src/main/java/org/eu/exodus_privacy/exodusprivacy/MainActivityViewModel.kt]: Ok, i removed this.

File: app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt [hunk 5]
```diff
@@ -11,6 +11,7 @@ import androidx.fragment.app.Fragment
 import androidx.fragment.app.viewModels
 import androidx.navigation.fragment.findNavController
 import androidx.recyclerview.widget.LinearLayoutManager
+import com.google.android.material.snackbar.Snackbar
```
Comment (Jean-BaptisteC) [inline app/src/main/java/org/eu/exodus_privacy/exodusprivacy/fragments/apps/AppsFragment.kt:14]:
```suggestion
```
Remove unused import

**Review by Jean-BaptisteC — CHANGES_REQUESTED** [review]
Please check if notification permission is granted and enable when user start refresh
Because i have deny permission and enable permission later, notification is never shown

**Comment by PatDyn** [comment]
> Please check if notification permission is granted and enable when user start refresh Because i have deny permission and enable permission later, notification is never shown

This happened, because the app would only create the notification channel during initial startup. When the user denies initially and decides otherwise later on, the notification channel is never created. According to: https://developer.android.com/develop/ui/views/notifications/channels#CreateChannel we could just create it unconditionally. 
I proposed this change in [f343b54](https://github.com/Exodus-Privacy/exodus-android-app/pull/273/commits/f343b54ea6f5188e324f25bde923092dde15712a).

**Review by Jean-BaptisteC — APPROVED** [review]
✅ Pixel 6 - Android 13
Tests ✅
